athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/GcsMetadataHandler.java

```diff
@@ -0,0 +1,471 @@
+     *
+     * @return An instance of {@link Schema}
+     */
+    public Schema getPartitionSchema()
+    {
+        SchemaBuilder schemaBuilder = SchemaBuilder.newBuilder()
+                .addField(BLOCK_PARTITION_COLUMN_NAME, Types.MinorType.VARCHAR.getType());
+        return schemaBuilder.build();
+    }
+
+    /**
+     * Used to get definition (field names, types, descriptions, etc...) of a Table.
+     *
+     * @param allocator Tool for creating and managing Apache Arrow Blocks.
+     * @param request   Provides details on who made the request and which Athena catalog, database, and table they are querying.
+     * @return A GetTableResponse which primarily contains:
+     * 1. An Apache Arrow Schema object describing the table's columns, types, and descriptions.
+     * 2. A Set<String> of partition column names (or empty if the table isn't partitioned).
+     * 3. A TableName object confirming the schema and table name the response is for.
+     * 4. A catalog name corresponding the Athena catalog that was queried.
+     */
+    @Override
+    public GetTableResponse doGetTable(BlockAllocator allocator, GetTableRequest request)
+    {
+        TableName tableInfo = request.getTableName();
+        LOGGER.debug("MetadataHandler=GcsMetadataHandler|Method=doGetTable|Message=queryId {}",
+                request.getQueryId());
+        if (this.datasource == null) {
+            throw new RuntimeException("Metastore couldn't be initialized");
+        }
+        LOGGER.debug("MetadataHandler=GcsMetadataHandler|Method=doGetTable|Message=Schema name {}, table name {}",
+                tableInfo.getSchemaName(), tableInfo.getTableName());
+        datasource.loadAllTables(tableInfo.getSchemaName());
+        LOGGER.debug(MessageFormat.format("Running doGetTable for table {0}, in schema {1} ",
+                tableInfo.getTableName(), tableInfo.getSchemaName()));
+        Schema schema = gcsSchemaUtils.buildTableSchema(this.datasource,
+                tableInfo.getSchemaName(),
+                tableInfo.getTableName());
+        Schema partitionSchema = getPartitionSchema();
+        return new GetTableResponse(request.getCatalogName(), request.getTableName(), schema,
+                partitionSchema.getFields().stream().map(Field::getName).collect(Collectors.toSet()));
+    }
+
+    /**
+     * Used to get the partitions that must be read from the request table in order to satisfy the requested predicate.
+     *
+     * @param blockWriter        Used to write rows (partitions) into the Apache Arrow response.
+     * @param request            Provides details of the catalog, database, and table being queried as well as any filter predicate.
+     * @param queryStatusChecker A QueryStatusChecker that you can use to stop doing work for a query that has already terminated
+     */
+    @Override
+    public void getPartitions(BlockWriter blockWriter, GetTableLayoutRequest request, QueryStatusChecker queryStatusChecker)
+    {
```

Comment:
At a high level, I'd like this partition/split architecture to follow a hive metastore format. That way, the knowledge people already have from storing their data in S3 for use with Athena can translate over. Similarly, the code would be similar.

https://sparkbyexamples.com/apache-hive/hive-partitions-explained-with-examples/ <- something like this

Reply:
I think it would make life easier for this implementation to be honest. Partitions are whatever the customer stores them as -- splits remains the same.

Reply:
TBD

Reply:
Hive-metastore-like partitioning is now place